src/connection/mqtt.rs

```diff
@@ -0,0 +1,619 @@
+        self.client
+            .publish(
+                format!(
+                    "{}/{}{}",
+                    self.client_id(),
+                    interface.interface_name(),
+                    path
+                ),
+                reliability,
+                false,
+                payload,
+            )
+            .await?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl<S> Connection<S> for Mqtt
+where
+    S: PropertyStore,
+{
+    type Payload = Bytes;
+
+    async fn connect(&self, device: &SharedDevice<S>) -> Result<(), crate::Error> {
+        loop {
+            match self.poll().await? {
+                rumqttc::Packet::ConnAck(connack) => {
+                    self.connack(device, connack).await?;
+
+                    return Ok(());
+                }
+                packet => warn!("Received incoming packet while waiting for connack: {packet:?}"),
+            }
+        }
+    }
+
+    async fn next_event(
+        &self,
+        device: &SharedDevice<S>,
+    ) -> Result<ReceivedEvent<Self::Payload>, crate::Error> {
+        static PURGE_PROPERTIES_TOPIC: OnceCell<String> = OnceCell::new();
+        static CLIENT_ID: OnceCell<String> = OnceCell::new();
+
+        // Keep consuming packets until we have an actual "data" event
+        loop {
+            match self.poll().await? {
+                rumqttc::Packet::ConnAck(connack) => self.connack(device, connack).await?,
+                rumqttc::Packet::Publish(publish) => {
+                    let purge_topic = PURGE_PROPERTIES_TOPIC.get_or_init(|| {
+                        format!("{}/control/consumer/properties", self.client_id())
+                    });
+
+                    debug!("Incoming publish = {} {:x}", publish.topic, publish.payload);
+
+                    if purge_topic == &publish.topic {
+                        debug!("Purging properties");
+
+                        self.purge_properties(device, &publish.payload).await?;
+                    } else {
+                        let client_id = CLIENT_ID.get_or_init(|| format!("{}", self.client_id()));
+                        let (interface, path) = parse_topic(client_id, &publish.topic)?;
+
+                        return Ok(ReceivedEvent {
+                            interface: interface.to_string(),
+                            path: path.to_string(),
+                            payload: publish.payload,
+                        });
+                    }
+                }
+                packet => {
+                    trace!("packet received {packet:?}");
+                }
+            }
+        }
+    }
+
+    fn deserialize_individual(
+        &self,
+        mapping: MappingRef<'_, &Interface>,
+        payload: &Self::Payload,
+    ) -> Result<(AstarteType, Option<Timestamp>), crate::Error> {
+        // TODO all validation that is independant from the transport should be moved outside and this function should just be called with a validated object
```

Comment:
Could we solve those todos with the new Serialize/Deserializ trait and a validated struct?

```rust
struct Validated<T>(T);

trait SerdePayload<P> {
 fn serialize_individual(...) -> Validated<AstarteType>;
 fn serialize_object(...) -> Validated<HashMap>;
 fn deser...
...
}
```

Reply:
The validated struct could maybe be created from a:

```rust
impl Validated<AstarteType> {
fn individual<P>(data: AstarteType) -> Result<Self>;
```

And internally call the shared validation on the return type T